
<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>gefes version 1.0.2 &mdash; gefes 1.0.2 documentation</title>
    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootstrap-3.3.6/css/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootstrap-3.3.6/css/bootstrap-theme.min.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/my-styles.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="_static/bootstrap-3.3.6/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="gefes 1.0.2 documentation" href="#" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body role="document">

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="#">
          gefes</a>
        <span class="navbar-text navbar-version pull-left"><b>1.0.2</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"></ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#"><code class="docutils literal"><span class="pre">gefes</span></code> version 1.0.2</a><ul>
<li><a class="reference internal" href="#warnings">Warnings</a></li>
<li><a class="reference internal" href="#context">Context</a></li>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#installing">Installing</a><ul>
<li><a class="reference internal" href="#step-1-cloning-the-repository">Step 1: Cloning the repository</a></li>
<li><a class="reference internal" href="#step-2a-modify-your-python-search-path">Step 2A: Modify your python search path</a></li>
<li><a class="reference internal" href="#step-2b-modify-your-binary-search-path">Step 2B: Modify your binary search path</a></li>
<li><a class="reference internal" href="#step-3-install-your-own-version-of-python">Step 3: Install your own version of python</a></li>
<li><a class="reference internal" href="#step-4-install-all-required-python-packages">Step 4: Install all required python packages</a></li>
<li><a class="reference internal" href="#step-5-obtaining-extra-dependencies">Step 5: Obtaining extra dependencies</a></li>
<li><a class="reference internal" href="#step-6-make-a-working-directory-with-the-raw-data-linked">Step 6: Make a working directory with the raw data linked</a></li>
<li><a class="reference internal" href="#step-7-make-a-json-file-describing-your-project">Step 7: Make a JSON file describing your project</a></li>
<li><a class="reference internal" href="#step-8-troubleshooting">Step 8: Troubleshooting</a></li>
</ul>
</li>
<li><a class="reference internal" href="#flowchart">Flowchart</a></li>
<li><a class="reference internal" href="#diagram">Diagram</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
              
            
            
            
            
              <li class="hidden-sm">
<div id="sourcelink">
  <a href="_sources/index.txt"
     rel="nofollow">Source</a>
</div></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  <div class="section" id="gefes-version-1-0-2">
<h1><code class="docutils literal"><span class="pre">gefes</span></code> version 1.0.2<a class="headerlink" href="#gefes-version-1-0-2" title="Permalink to this headline">¶</a></h1>
<p>The acronym <code class="docutils literal"><span class="pre">gefes</span></code> stands for <strong>G</strong>​enome <strong>E</strong>​xtraction
<strong>F</strong>​rom <strong>E</strong>​nvironmental <strong>S</strong>​equencing.</p>
<p>This is yet another pipeline for assembling the short reads produced by
shotgun-metagenomic sequencing experiments in an attempt to recompose
full microbial genomes. With this tool, we would like to reconstitute
the functional potential of the important bacterial and archaeal players
in aquatic environments.</p>
<p>This source code is propriety of Lucas Sinclair <a class="reference external" href="mailto:lucas&#37;&#52;&#48;sinclair&#46;bio">lucas<span>&#64;</span>sinclair<span>&#46;</span>bio</a>, a
bioinformatics consultant hireable at www.sinclair.bio</p>
<div class="section" id="warnings">
<h2>Warnings<a class="headerlink" href="#warnings" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>You also have to keep in mind that what we are measuring is a
potential. It&#8217;s not because you see a particular gene in a population
or strain that it is necessarily being actively translated,
transcribed, folded and exported. Ideally, you should use the results
from such tools like <code class="docutils literal"><span class="pre">gefes</span></code> to direct your laboratory assays which
will truly confirm if a given process is taking place.</li>
<li>The <code class="docutils literal"><span class="pre">gefes</span></code> project is not a biologist-oriented tool that supports
all the possible use cases one could have with metagenomic sequence
data out of the box. For instance, it does not have a graphical
interface to operate, nor any bash/sh/csh commands. Indeed, as each
sequencing experiment will have different goals and scientific
questions associated to it, there cannot be a standard set of
procedures to apply to the data. Instead, the <code class="docutils literal"><span class="pre">gefes</span></code> project a
flexible and modular collections of packages written in proper, clean
and commented object-oriented python which enables the user to
survey, modify and extend the code-base easily &#8211; provided he has a
sufficient knowledge in programming. It is a basis upon which the
scientist can set up the processing and analysis that he sees fit for
his own data sparing him from having to develop lots of the
infrastructure needed himself.</li>
</ul>
</div>
<div class="section" id="context">
<h2>Context<a class="headerlink" href="#context" title="Permalink to this headline">¶</a></h2>
<p>As you know, almost none of the microbes living in natural environments
can be isolated or cultured easily. So instead, we go and perform
shotgun metagenomic sampling by taking a glass of water from a lake,
operating a total DNA extraction and inserting the solution into a
high-throughput sequencer. As a result, we receive a file full of short
DNA reads each coming (statistically) from a different microbe.</p>
<p>It&#8217;s quite different from when you are able to isolate a bacterium and
grow it such as E. Coli. In that case, your DNA reads are coming from
random locations of the E. Coli. chromosome, but they are all
originating from a copy of the same genome. This makes it easy to pieces
things together afterwards.</p>
<p>What we have as starting data in our case is more messy. Every DNA read
is potentially coming from a different species. Plus, the fragments are
really short and only span a fraction of a typical microbial gene.</p>
<p>How do we put the reads together to make genomes ? How are we going to
figure out which short sequence was coming from which species ? This is
what <code class="docutils literal"><span class="pre">gefes</span></code> is supposed to help with.</p>
<p>Many objects common to any analysis such as a &#8220;FASTQ file pair&#8221;, a
&#8220;Sample&#8221;, a &#8220;Aggregate of Samples&#8221;, a &#8220;Sequence quality checker&#8221;, an
&#8220;Assembly&#8221;, a &#8220;Read mapper&#8221;, a &#8220;Contig binner&#8221;, etc. are provided. In
addition you will find routines for sending these objects through
well-known algorithms such as Sickle, Ray, Bowtie, etc. Lots of other
functionality is also present such as a multitude of visualizations in
<code class="docutils literal"><span class="pre">matplotlib</span></code> and other things such as the ability to automatically
distribute the computation on a network of computers (via the SLURM
queuing system).</p>
</div>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>Starting from the raw reads there are about eight distinct processing
steps in <code class="docutils literal"><span class="pre">gefes</span></code>:</p>
<ol class="arabic simple">
<li>Clean</li>
<li>Co-Assemble</li>
<li>Map</li>
<li>Bin contigs together</li>
<li>Validate</li>
<li>Assign taxonomy</li>
<li>Annotate</li>
<li>Predict metabolism</li>
</ol>
<p>Unfortunately, no other detailed documentation has been written yet but
the code is clean and commented. In addition, these two descriptive
files might help you figure out what is going on:</p>
<ul class="simple">
<li><a class="reference external" href="/../master/documentation/flowchart.pdf?raw=true">Flowchart of data
processing</a></li>
<li><a class="reference external" href="/../master/documentation/diagram.pdf?raw=true">Tentative UML diagram of objects
composition</a></li>
</ul>
</div>
<div class="section" id="installing">
<h2>Installing<a class="headerlink" href="#installing" title="Permalink to this headline">¶</a></h2>
<p>No automated installation has been developed for the <code class="docutils literal"><span class="pre">gefes</span></code> package
yet. In the meantime, following this document and typing these commands
on your bash prompt should get you started. It is designed so you don&#8217;t
need super user privileges at any step. If you cannot get a functional
installation set up, contact the authors.</p>
<div class="section" id="step-1-cloning-the-repository">
<h3>Step 1: Cloning the repository<a class="headerlink" href="#step-1-cloning-the-repository" title="Permalink to this headline">¶</a></h3>
<p>Here you will download a copy of the code from github and place it in
your home directory.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ cd ~
$ mkdir repos
$ cd repos
$ git clone git@github.com:xapple/gefes.git
</pre></div>
</div>
<p>NB: the access to this repository is restricted.</p>
</div>
<div class="section" id="step-2a-modify-your-python-search-path">
<h3>Step 2A: Modify your python search path<a class="headerlink" href="#step-2a-modify-your-python-search-path" title="Permalink to this headline">¶</a></h3>
<p>Here you will edit your <code class="docutils literal"><span class="pre">.bashrc</span></code> or <code class="docutils literal"><span class="pre">.bash_profile</span></code> to add a
reference to the module you just downloaded. When you type
<code class="docutils literal"><span class="pre">import</span> <span class="pre">gefes</span></code> python will know where to look !</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ vim ~/.bash_profile
export PYTHONPATH=&quot;$HOME/repos/gefes/&quot;:$PYTHONPATH
</pre></div>
</div>
</div>
<div class="section" id="step-2b-modify-your-binary-search-path">
<h3>Step 2B: Modify your binary search path<a class="headerlink" href="#step-2b-modify-your-binary-search-path" title="Permalink to this headline">¶</a></h3>
<p>There are some dependencies that are not too large and can be bundled as
binaries with this module. To avoid the hassle of downloading and
compiling these requirements, we have added a bunch of them in the
repository. Add them to your <code class="docutils literal"><span class="pre">$PATH</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ vim ~/.bash_profile
export PATH=&quot;$HOME/repos/gefes/bin/linux/:$PATH&quot;
export PATH=&quot;$HOME/repos/gefes/bin/linux/mcl/:$PATH&quot;
</pre></div>
</div>
</div>
<div class="section" id="step-3-install-your-own-version-of-python">
<h3>Step 3: Install your own version of python<a class="headerlink" href="#step-3-install-your-own-version-of-python" title="Permalink to this headline">¶</a></h3>
<p>Your system probably comes with a version of python installed. But the
variations from system to system are too great to rely on any available
setup. We strongly recommand to just install our own in the home
directory. Also, we will then be able to install modules without
administrator privileges. You can skip this step if you are confident
enough about what you are doing.</p>
<p>For this we will be using this excellent project:
<a class="reference external" href="https://github.com/yyuu/pyenv">https://github.com/yyuu/pyenv</a></p>
<p>To install it you may use this sister project:
<a class="reference external" href="https://github.com/yyuu/pyenv-installer">https://github.com/yyuu/pyenv-installer</a></p>
<p>Basically you just need to type this command:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ curl -L https://raw.githubusercontent.com/yyuu/pyenv-installer/master/bin/pyenv-installer | bash
</pre></div>
</div>
<p>These lines go into your <code class="docutils literal"><span class="pre">.bash_profile</span></code>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ vim ~/.bash_profile
export PYENV_ROOT=&quot;$HOME/.pyenv&quot;
export PATH=&quot;$PYENV_ROOT/bin:$PATH&quot;
eval &quot;$(pyenv init -)&quot;
</pre></div>
</div>
<p>Relaunch your shell and type these commands to get the right version of
python:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ pyenv install 2.7.11
$ pyenv global 2.7.11
$ pyenv rehash
</pre></div>
</div>
</div>
<div class="section" id="step-4-install-all-required-python-packages">
<h3>Step 4: Install all required python packages<a class="headerlink" href="#step-4-install-all-required-python-packages" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">gefes</span></code> uses many third party python libraries. You can get them by
running these commands:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ pip install --user matplotlib
$ pip install --user pandas
$ pip install --user biopython
$ pip install --user sh
$ pip install --user tqdm
$ pip install --user decorator
$ pip install --user shell_command
$ pip install --user threadpool
$ pip install --user scipy
$ pip install --user pysam
$ pip install --user concoct
$ pip install --user tabulate
$ pip install --user regex
$ pip install --user humanfriendly
$ pip install --user ftputil
$ pip install --user pystache
$ pip install --user humanfriendly
$ pip install --user humanfriendly
$ pip install --user humanfriendly
$ pip install --user humanfriendly
$ pip install --user humanfriendly
</pre></div>
</div>
<p>It also uses several first-party python libraries that we have developed
and use in several projects. You can get them by running these commands:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ pip install plumbing
$ pip install fasta
$ pip install pymarktex
</pre></div>
</div>
<p>Don&#8217;t forget to rehash the executable links at the end if you are using
<code class="docutils literal"><span class="pre">pyenv</span></code> like we do:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ pyenv rehash
</pre></div>
</div>
</div>
<div class="section" id="step-5-obtaining-extra-dependencies">
<h3>Step 5: Obtaining extra dependencies<a class="headerlink" href="#step-5-obtaining-extra-dependencies" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">gefes</span></code> makes use of many third party programs which need to be
installed and accessible from your <code class="docutils literal"><span class="pre">$PATH</span></code>. Depending on what parts of
the pipeline you are planning to run, you might not need them all. You
can try and install the missing ones only when <code class="docutils literal"><span class="pre">gefes</span></code> complains about
a missing executable. These dependencies each have specific installation
procedures and include:</p>
<ul class="simple">
<li><a class="reference external" href="http://blast.be-md.ncbi.nlm.nih.gov/Blast.cgi">NCBI BLAST</a>
version 2.2.28+ providing <code class="docutils literal"><span class="pre">blastn</span></code> and <code class="docutils literal"><span class="pre">blastp</span></code></li>
<li><a class="reference external" href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">bowtie2</a>
version 2.2.4 providing <code class="docutils literal"><span class="pre">bowtie2</span></code> and <code class="docutils literal"><span class="pre">bowtie2-build</span></code></li>
<li><a class="reference external" href="http://bedtools.readthedocs.org/en/latest/">bedtools</a> version
2.15.0 providing <code class="docutils literal"><span class="pre">genomeCoverageBed</span></code></li>
<li><a class="reference external" href="http://samtools.sourceforge.net">samtools</a> version 0.1.19
providing <code class="docutils literal"><span class="pre">samtools</span></code></li>
<li><a class="reference external" href="http://www.vicbioinformatics.com/software.prokka.shtml">prokka</a>
version 1.10 providing <code class="docutils literal"><span class="pre">prokka</span></code> (itself requires rnammer, aragorn,
prodigal, barrnap, signalp and tbl2asn)</li>
<li><a class="reference external" href="https://www.tug.org/texlive/">LaTeX</a> version <code class="docutils literal"><span class="pre">TeX</span> <span class="pre">Live</span> <span class="pre">2014</span></code>
providing <code class="docutils literal"><span class="pre">xelatex</span></code>.</li>
<li><a class="reference external" href="http://pandoc.org/">pandoc</a> version <code class="docutils literal"><span class="pre">1.17.0.3</span></code> providing
<code class="docutils literal"><span class="pre">pandoc</span></code>.</li>
</ul>
<p>These can take some time to install and unfortunately we can&#8217;t package
them with our project! Hopefully, some of them are already installed on
your server or can be accessed via a module system. Then, there are a
few that are bundled within this repository or are obtained with
<code class="docutils literal"><span class="pre">pip</span></code>:</p>
<ul class="simple">
<li><a class="reference external" href="http://sourceforge.net/projects/denovoassembler/">Ray</a> version
2.3.1 providing <code class="docutils literal"><span class="pre">ray231</span></code> (included in repository)</li>
<li><a class="reference external" href="http://broadinstitute.github.io/picard/">Picard Tools</a> version
1.101 providing <code class="docutils literal"><span class="pre">MarkDuplicates.jar</span></code> (included in repository)</li>
<li><a class="reference external" href="https://github.com/BinPro/CONCOCT">concoct</a> version 0.4.0
providing <code class="docutils literal"><span class="pre">concoct</span></code> (included in the python package index)</li>
</ul>
</div>
<div class="section" id="step-6-make-a-working-directory-with-the-raw-data-linked">
<h3>Step 6: Make a working directory with the raw data linked<a class="headerlink" href="#step-6-make-a-working-directory-with-the-raw-data-linked" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">gefes</span></code> will generate results in a directory named <code class="docutils literal"><span class="pre">views</span></code> in a
directory named <code class="docutils literal"><span class="pre">GEFES</span></code> in your home folder. You can change this by
editing the code of course.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ cd ~
$ mkdir GEFES
$ cd GEFES
$ mkdir views
</pre></div>
</div>
</div>
<div class="section" id="step-7-make-a-json-file-describing-your-project">
<h3>Step 7: Make a JSON file describing your project<a class="headerlink" href="#step-7-make-a-json-file-describing-your-project" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">gefes</span></code> will create <code class="docutils literal"><span class="pre">Project</span></code> objects with associated <code class="docutils literal"><span class="pre">Sample</span></code>
objects based on a user inputted JSON files. Look at the <code class="docutils literal"><span class="pre">json</span></code>
directory at the root of the repository and make your own. You have to
tell <code class="docutils literal"><span class="pre">gefes</span></code> where to look for your raw reads.</p>
</div>
<div class="section" id="step-8-troubleshooting">
<h3>Step 8: Troubleshooting<a class="headerlink" href="#step-8-troubleshooting" title="Permalink to this headline">¶</a></h3>
<p>You might run into difficulties at some of the steps above &#8211; if that is
the case check this section for a solution.</p>
<ul class="simple">
<li>&#8220;pip install scipy&#8221; missing BLAS:</li>
</ul>
<hr class="docutils" />
<p>When you install scipy, you might need these two dependencies before
hand if they are not on your system:
<a class="reference external" href="http://stackoverflow.com/questions/7496547/python-scipy-needs-blas">http://stackoverflow.com/questions/7496547/python-scipy-needs-blas</a></p>
<ul class="simple">
<li>git version too old:</li>
</ul>
<hr class="docutils" />
<p>If your git version is too old you can follow these steps:
<a class="reference external" href="http://blog.justin.kelly.org.au/install-git-in-your-home-directory-how-to/">http://blog.justin.kelly.org.au/install-git-in-your-home-directory-how-to/</a></p>
<ul class="simple">
<li>Compiling Ray:</li>
</ul>
<hr class="docutils" />
<p>To compile Ray on a compute cluster you might have to do something this:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ module swap PrgEnv-intel PrgEnv-gnu
$ make -j24 MPI_IO=y MPICXX=CC MAXKMERLENGTH=91
</pre></div>
</div>
<p>Or something like this:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ module load openmpi/1.4.5
$ module load pgi
$ make -j16 MPI_IO=y MPICXX=mpicc MAXKMERLENGTH=91
</pre></div>
</div>
<ul class="simple">
<li>&#8220;pip install matplotlib&#8221; missing freetype:</li>
</ul>
<hr class="docutils" />
<p>If you are on OS X you can simply fix this error by typing
<code class="docutils literal"><span class="pre">$</span> <span class="pre">brew</span> <span class="pre">install</span> <span class="pre">freetype</span></code></p>
</div>
</div>
<div class="section" id="flowchart">
<h2>Flowchart<a class="headerlink" href="#flowchart" title="Permalink to this headline">¶</a></h2>
<p>Below is drawn the flowchart describing the data processing along all
the steps of GEFES:</p>
<div class="figure" id="id1">
<img alt="Flowchart" src="_images/flowchart.png" />
<p class="caption"><span class="caption-text">Flowchart</span></p>
</div>
</div>
<div class="section" id="diagram">
<h2>Diagram<a class="headerlink" href="#diagram" title="Permalink to this headline">¶</a></h2>
<p>Below is presented a tentative UML diagram detailing not the inheritance
of the classes but the use of the composition design pattern between the
objects created.</p>
<div class="figure" id="id2">
<img alt="Diagram" src="_images/diagram.png" />
<p class="caption"><span class="caption-text">Diagram</span></p>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2016, Lucas Sinclair.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.<br/>
    </p>
  </div>
</footer>
  </body>
</html>